rpcs3/Emu/Cell/SPULLVMRecompiler.cpp

```diff
@@ -5980,39 +6007,208 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 			{
-				if (auto [ok_fm1, a3, b3] = match_expr(c, fm(MT, MT)); ok_fm1 && a3.eq(a1))
+				if (auto [ok_fm2, fm_half_mul] = match_fm_half(); ok_fm2 && fm_half_mul.eq(b1))
 				{
-					if (auto [ok_sqrte, src] = match_expr(a3, spu_rsqrte(MT)); ok_sqrte && src.eq(b3))
+					if (fm_half_mul.eq(b1))
 					{
-						erase_stores(a, b, c, a3);
-						set_vr(op.rt4, fsqrt(fabs(src)));
-						return;
+						if (auto [ok_fm1, a3, b3] = match_expr(c, fm(MT, MT)); ok_fm1 && a3.eq(a1))
+						{
+							if (auto [ok_sqrte, src] = match_expr(a3, spu_rsqrte(MT)); ok_sqrte && src.eq(b3))
+							{
+								erase_stores(a, b, c, a3);
+								set_vr(op.rt4, fsqrt(fabs(src)));
+								return true;
+							}	
+						}
+						else if (auto [ok_fm1, a3, b3] = match_expr(c, fm(MT, MT)); ok_fm1 && b3.eq(a1))
+						{
+							if (auto [ok_sqrte, src] = match_expr(b3, spu_rsqrte(MT)); ok_sqrte && src.eq(a3))
+							{
+								erase_stores(a, b, c, b3);
+								set_vr(op.rt4, fsqrt(fabs(src)));
+								return true;
+							}	
+						}
+					}
+					else if (fm_half_mul.eq(a1))
+					{
+						if (auto [ok_fm1, a3, b3] = match_expr(c, fm(MT, MT)); ok_fm1 && a3.eq(b1))
+						{
+							if (auto [ok_sqrte, src] = match_expr(a3, spu_rsqrte(MT)); ok_sqrte && src.eq(b3))
+							{
+								erase_stores(a, b, c, a3);
+								set_vr(op.rt4, fsqrt(fabs(src)));
+								return true;
+							}	
+						}
+						else if (auto [ok_fm1, a3, b3] = match_expr(c, fm(MT, MT)); ok_fm1 && b3.eq(b1))
+						{
+							if (auto [ok_sqrte, src] = match_expr(b3, spu_rsqrte(MT)); ok_sqrte && src.eq(a3))
+							{
+								erase_stores(a, b, c, b3);
+								set_vr(op.rt4, fsqrt(fabs(src)));
+								return true;
+							}	
+						}
 					}
 				}
 			}
-		}
 
-		// Match division (fast)
-		if (auto [ok_fnma, divb, diva] = match_expr(a, fnms(c, MT, MT)); ok_fnma)
+			return false;
+		};
+
+		if (check_sqrt_pattern_for_float(1.0f))
+			return;
+
+		if (check_sqrt_pattern_for_float(std::bit_cast<f32>(std::bit_cast<u32>(1.0f) + 1)))
+			return;
+
+		auto check_accurate_reciprocal_pattern_for_float = [&](f32 float_value) -> bool
 		{
-			if (auto [ok_fm] = match_expr(c, fm(diva, b)); ok_fm)
+			// FMA(FNMS(div, spu_re(div), float_value), spu_re(div), spu_re(div))
+			if (auto [ok_fnms, div] = match_expr(a, fnms(MT, b, fsplat<f32[4]>(float_value))); ok_fnms && op.rb == op.rc)
 			{
-				if (auto [ok_re] = match_expr(b, spu_re(divb)); ok_re)
+				if (auto [ok_re] = match_expr(b, spu_re(div)); ok_re)
 				{
-					erase_stores(b, c);
-					set_vr(op.rt4, diva / divb);
-					return;
+					erase_stores(b);
+					set_vr(op.rt4, re_accurate(div));
+					return true;
+				}
+			}
+
+			// FMA(FNMS(spu_re(div), div, float_value), spu_re(div), spu_re(div))
+			if (auto [ok_fnms, div] = match_expr(a, fnms(b, MT, fsplat<f32[4]>(float_value))); ok_fnms && op.rb == op.rc)
+			{
+				if (auto [ok_re] = match_expr(b, spu_re(div)); ok_re)
+				{
+					erase_stores(b);
```

Comment:
I think, you are missing a check that checks of the FNMS addend is the same as FMA